### PR TITLE
Move lib/ events to IEventDispatcher where possible

### PR DESCRIPTION
**Pull request by nickvergessen**
* Another step towards #38546
* Based on #39485

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Comment by st3iny** [comment]
Code is too deep for me. PHP CS check is unhappy though.

File: lib/private/Share20/LegacyHooks.php [hunk 1]
```diff
@@ -26,35 +26,50 @@
...
+				$this->postShare($event);
+			}
+		});
 	}
 
 	/**
 	 * @param GenericEvent $e
 	 */
-	public function preUnshare(GenericEvent $e) {
+	public function preUnshare($e) {
```
Comment (come-nc) [inline lib/private/Share20/LegacyHooks.php:72]: Why is the typing removed? The call is wrapped in the `if ($event instanceof GenericEvent) {` so it should be good?

**Review by come-nc — COMMENTED** [review]
Did you keep track of all removed legacy events to fill the upgrade documentation accordingly?

File: lib/private/Share20/LegacyHooks.php (hunk 1 above)
Comment (nickvergessen) [reply lib/private/Share20/LegacyHooks.php:72]: Right, seems to be from an intermediate step.

But as per https://github.com/nextcloud/server/pull/39485#discussion_r1275052429 it also doesn't seem to matter too much, as mostlikely we will have to remove GenericEvent as well (assuming symfony 6.4 will be required for PHP 8.3 support).
But can readd it for now.

**Comment by nickvergessen** [comment]
> Did you keep track of all removed legacy events to fill the upgrade documentation accordingly?

Yes, will send a PR once the last remaining events are migrated. Todo list is:
https://github.com/nextcloud/server/issues?q=label%3A%22pending+documentation%22+author%3Anickvergessen+is%3Aclosed 😎

**Review by come-nc — APPROVED** [review]

**Review by ChristophWurst — APPROVED** [review]

**Comment by nickvergessen** [comment]
Docs in https://github.com/nextcloud/documentation/pull/10858

**Pull request merged**